File: 2018/07-the-sum-of-its-parts/rust/src/main.rs
use petgraph::{Graph, Incoming};
use petgraph::graph::NodeIndex;
use std::collections::HashMap;

/*
const PUZZLE: &str = "
Step C must be finished before step A can begin.
Step C must be finished before step F can begin.
Step A must be finished before step B can begin.
Step A must be finished before step D can begin.
Step B must be finished before step E can begin.
Step D must be finished before step E can begin.
Step F must be finished before step E can begin.
";
const PUZZLE: &str = include_str!("../../input");
*/
const PUZZLE: &str = "
Step A must be finished before step B can begin.
Step B must be finished before step C can begin.
Step C must be finished before step D can begin.
Step D must be finished before step E can begin.
Step E must be finished before step F can begin.
Step F must be finished before step G can begin.
Step G must be finished before step H can begin.
";

fn main() {
    let graph = make_graph(PUZZLE);
    println!("A: {}", solve_a(&graph));
    println!("B: {}", solve_b(&graph));
}

// read in the nodes from the source input text
fn make_graph(input: &str) -> Graph<char, char> {
    let mut graph = Graph::<char, char>::new();
    let mut nodes = HashMap::new();

    // collect up all the instructions
    // fortunately the input is well formed: single letters at exact positions
    let instructions: Vec<(char, char)> = input.trim().lines()
        .filter(|x| x.contains("Step"))
        .map(|x| (x.chars().nth(5).unwrap(), x.chars().nth(36).unwrap()))
        .collect()
        ;

    // create the nodes
    for i in instructions.iter() {
        if let None = nodes.get(&i.0) {
            let index = graph.add_node(i.0);
            nodes.insert(&i.0, index);
        }
        if let None = nodes.get(&i.1) {
            let index = graph.add_node(i.1);
            nodes.insert(&i.1, index);
        }
    }

    // create the edges/lines
    for i in instructions.iter() {
        let n1 = nodes.get(&i.0).unwrap();
        let n2 = nodes.get(&i.1).unwrap();
        graph.add_edge(*n1, *n2, ' ');
    }

    graph
}

fn solve_a(graph: &Graph<char, char>) -> String {
    // start with all nodes without any incomming neighbors
    let mut available = Vec::new();
    for n in graph.node_indices() {
        if graph.neighbors_directed(n, Incoming).count() == 0 {
            available.push((graph[n], n));
        }
    }

    let mut path = Vec::new();

    while available.len() > 0 {
        // sort the list into alphabetical order so when we are picking the
        // next edge to follow it is the lowest alphabetically
        available.sort_unstable_by(|a, b| a.cmp(b));

        let mut chosen_index = None;

        // determine which next edge we can follow based on dependancies
        for (i, a) in available.iter().enumerate() {
            let n_count = graph.neighbors_directed(a.1, Incoming).count();
            let d_count = graph.neighbors_directed(a.1, Incoming)
                .filter(|x| path.contains(x))
                .count();

            if (n_count - d_count) == 0 {
                chosen_index = Some(i);
                break;
            }
        }

        if chosen_index == None {
            break;
        }

        let node = available[chosen_index.unwrap()].1;

        available.remove(chosen_index.unwrap());

        // add this node to the list we have process
        path.push(node);

        // keep track of which edges are now available paths to take
        for n in graph.neighbors(node) {
            let element = (graph[n], n);
            if ! available.contains(&element) {
                available.push(element);
            }
        }
    }

    path.iter().map(|x| graph[*x]).collect::<String>()
}

struct Worker {
    id: u8,
    job: Option<NodeIndex>,
    time_remaining: usize,
}

impl Worker {
    fn new(id: u8) -> Worker {
        Worker {
            id: id,
            job: None,
            time_remaining: 0,
        }
    }

    fn is_finished(&self) -> bool {
        self.job != None && self.time_remaining == 0
    }

    fn is_working(&self) -> bool {
        self.job != None && self.time_remaining > 0
    }

    fn tick(&mut self) {
        self.time_remaining -= 1;
    }

    fn work_on(&mut self, job: NodeIndex, cost: usize) {
        self.job = Some(job);
        self.time_remaining = cost;
    }

    // reset the worker to an initial state, returns previously worked on job
    // will panic if reset when not working on anything
    fn reset(&mut self) -> NodeIndex {
        let completed_unit = self.job.unwrap();
        self.job = None;
        self.time_remaining = 0;
        completed_unit
    }
}

fn solve_b(graph: &Graph<char,char>) -> usize {
    //const ALPHABET_SECONDS: usize = 1_560; // fixed time we must add, 60 * 26

    // our 5 workers ready to build
    let mut workers = Vec::new();
    for id in 0u8..5 {
        workers.push(Worker::new(id));
    }

    // start with all nodes without any incomming neighbors
    // there are more than one in the puzzle input so a few workers will start
    // building at the same time
    let mut available = Vec::new();
    for n in graph.node_indices() {
        if graph.neighbors_directed(n, Incoming).count() == 0 {
            available.push((graph[n], n));
        }
    }

    // though we know the path from solve_a we can use it again to help us
    // track dependancies
    let mut path = Vec::new();

    // how many seconds have pass, part of the result
    let mut ticks = 0;

    while path.len() < graph.node_count() {
        println!("tick: {} path: {} nodes: {}", ticks, path.len(), graph.node_count());
        // sort the list into alphabetical order so when we are picking the
        // next edge to follow it is the lowest alphabetically
        available.sort_unstable_by(|a, b| a.cmp(b));

        // handle each worker
        for worker in &mut workers {
            print!("worker[{}]", worker.id);
            if worker.is_working() {
                println!(" is working on {} for {} more ticks", graph[worker.job.unwrap()], worker.time_remaining);
                worker.tick();
                continue;
            } else if worker.is_finished() {
                print!(" is finsihed with {}", graph[worker.job.unwrap()]);
                path.push(worker.reset());
                // this worker can work on something new this tick
            }
            print!(" is available for work");

            // if we get here then the worker is available to work

            let mut chosen_index = None;

            // determine which next edge we can follow based on dependancies
            for (i, a) in available.iter().enumerate() {
                let n_count = graph.neighbors_directed(a.1, Incoming).count();
                let d_count = graph.neighbors_directed(a.1, Incoming)
                    .filter(|x| path.contains(x))
                    .count();

                if (n_count - d_count) == 0 {
                    chosen_index = Some(i);
                    break;
                }
            }

            if chosen_index == None {
                println!(" but no work is available right now!");
                continue;
            }

            let node = available[chosen_index.unwrap()].1;

            available.remove(chosen_index.unwrap());

            let node_cost = ((graph[node] as u8 - 65) as usize) + 60;

            println!(" and will work on {} for {} ticks", graph[node], node_cost);

            worker.work_on(node, node_cost);

            // keep track of which edges are now available paths to take
            for n in graph.neighbors(node) {
                let element = (graph[n], n);
                if ! available.contains(&element) {
                    available.push(element);
                }
            }
        }

        ticks += 1;
    }

    //ticks + ALPHABET_SECONDS
    ticks - 1
}

File: 2018/07-the-sum-of-its-parts/main.go
package main

import (
	_ "embed"
	"flag"
	"fmt"
	"log"
	"os"
	"runtime/pprof"
	"sort"
	"strings"
)

//go:embed input.txt
var input string

var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")

func main() {
	flag.Parse()
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	lines := parseInput(input)

	listA := newAdjacencyList(lines)
	fmt.Println("A:", SolveA(listA))

	listB := newAdjacencyList(lines)
	fmt.Println("B:", SolveB(listB, 5, 60))
}

func parseInput(s string) []string {
	return strings.Split(strings.TrimSpace(s), "\n")
}

type AdjacencyItem struct {
	outgoing []rune
	incoming []rune
	isActive bool // For Part 2
}

type AdjacencyList map[rune]AdjacencyItem

func newAdjacencyList(lines []string) AdjacencyList {
	list := make(AdjacencyList)
	for _, line := range lines {
		// Step B must be finished before step E can begin.
		keyNode := rune(line[5])
		edgeNode := rune(line[36])
		// keyNode := rune(line[5] - 65)
		// edgeNode := rune(line[36] - 65)

		if entry, ok := list[keyNode]; ok {
			entry.outgoing = append(entry.outgoing, edgeNode)
			list[keyNode] = entry
		} else {
			entry := AdjacencyItem{outgoing: make([]rune, 1, 26), incoming: make([]rune, 0, 26)}
			entry.outgoing[0] = edgeNode
			list[keyNode] = entry
		}
	}

	for keyNode := range list {
		for _, outgoingNode := range list[keyNode].outgoing {
			if entry, ok := list[outgoingNode]; ok {
				entry.incoming = append(entry.incoming, keyNode)
				list[outgoingNode] = entry
			} else {
				entry := AdjacencyItem{outgoing: []rune{}, incoming: make([]rune, 1, 26)}
				entry.incoming[0] = keyNode
				list[outgoingNode] = entry
			}
		}
	}

	return list
}

/*
func (list AdjacencyList) SortedKeys() []rune {
	sortedKeys := make([]rune, 0, len(list))
	for key := range list {
		sortedKeys = append(sortedKeys, key)
	}
	sort.Slice(sortedKeys, func(i, j int) bool {
		return sortedKeys[i] < sortedKeys[j]
	})
	return sortedKeys
}
*/

func (list AdjacencyList) AvailableKeys() []rune {
	// all keys without incoming neighbors
	// set the initial capacity to the number of keys
	// it probably won't reach that, but won't be any longer
	availableKeys := make([]rune, 0, len(list))
	for key := range list {
		if len(list[key].incoming) == 0 {
			availableKeys = append(availableKeys, key)
		}
	}
	sort.Slice(availableKeys, func(i, j int) bool { return availableKeys[i] > availableKeys[j] })
	return availableKeys
}

/*
func removeNodeFromSlice(s []rune, needle rune) []rune {
	newSlice := make([]rune, 0, len(s)-1)
	for i := 0; i < len(s); i++ {
		if s[i] != needle {
			newSlice = append(newSlice, s[i])
		}
	}
	return newSlice
}
*/

func removeNodeFromSlice(s []rune, needle rune) []rune {
	found := false
	for i, j := 0, 0; i < len(s); i++ {
		if !found && s[i] != needle {
			found = true
			continue
		}
		if found {
			s[j] = s[i]
		}
		j++
	}
	return s[0 : len(s)-1]
}

func (list AdjacencyList) RemoveNode(key rune) {
	// remove the node and any references to it
	incomingNodes := list[key].incoming
	outgoingNodes := list[key].outgoing
	delete(list, key)

	for _, incomingKey := range incomingNodes {
		entry := list[incomingKey]
		entry.outgoing = removeNodeFromSlice(entry.outgoing, key)
		list[incomingKey] = entry
	}

	for _, outgoingKey := range outgoingNodes {
		entry := list[outgoingKey]
		entry.incoming = removeNodeFromSlice(entry.incoming, key)
		list[outgoingKey] = entry
	}
}

func SolveA(list AdjacencyList) string {
	availableKeys := list.AvailableKeys()

	path := make([]rune, 0, len(list))
	for len(availableKeys) > 0 {
		// find the next node to follow
		// first node without any dependancies
		chosenKey := rune(-1)
		for _, key := range availableKeys {
			if len(list[key].incoming) == 0 {
				chosenKey = key
			}
		}
		if chosenKey < 0 {
			break
		}

		path = append(path, chosenKey)

		list.RemoveNode(chosenKey)

		availableKeys = list.AvailableKeys()
	}
	return string(path)
}

type Worker struct {
	node          rune
	timeRemaining int
}

func (w *Worker) HasJob() bool {
	return w.node != 0
}

func (w *Worker) IsFinished() bool {
	return w.node != 0 && w.timeRemaining == 0
}

func (w *Worker) WorkOn(node rune, ticksPerJob int) {
	w.node = node
	w.timeRemaining = int(node) - 65 + ticksPerJob
}

func (w *Worker) Tick() {
	w.timeRemaining--
}

func (w *Worker) Reset() {
	w.node = 0
}

func SolveB(list AdjacencyList, totalWorkers int, ticksPerJob int) int {
	workers := make([]Worker, totalWorkers, totalWorkers)
	for i := 0; i < totalWorkers; i++ {
		workers[i] = Worker{}
	}

	completedJobs := 0
	totalJobs := len(list)

	availableKeys := list.AvailableKeys()

	ticks := -1
	for completedJobs < totalJobs {
		// fmt.Printf("tick: %d completed: %d nodes: %d available: %d workers: %#v\n", ticks, completedJobs, totalJobs, len(availableKeys), workers)
		for workerIndex := range workers {
			// fmt.Printf("worker[%d] contents: %v ", workerIndex, workers[workerIndex].node)
			if workers[workerIndex].HasJob() { // is active
				if !workers[workerIndex].IsFinished() { // still working
					// fmt.Printf(" is working on %s for %d more ticks\n", string(workers[workerIndex].node), workers[workerIndex].timeRemaining)
					workers[workerIndex].Tick()
					continue
				} else { // finished
					// fmt.Printf(" is finished with %s ", string(workers[workerIndex].node))
					list.RemoveNode(workers[workerIndex].node)
					availableKeys = list.AvailableKeys()

					workers[workerIndex].Reset()

					completedJobs++
					if completedJobs == totalJobs {
						// fmt.Printf("(ALL JOBS COMPLETE!)\n")
						break
					}
				}
			}
			// fmt.Printf("is available for work ")

			// the worker is ready for a new job...

			// find the next node to follow
			// first node without any dependancies
			// and not being worked on
			chosenKey := rune(-1)
			for _, key := range availableKeys {
				item := list[key]
				if len(item.incoming) == 0 && !item.isActive {
					chosenKey = key
				}
			}
			if chosenKey < 0 {
				// no work available yet
				// process other workers until work becomes available again
				// fmt.Printf("but no work is available right now!\n")
				continue
			}

			// the worker is ready for a job and there is work available...

			entry := list[chosenKey]
			entry.isActive = true
			list[chosenKey] = entry

			workers[workerIndex].WorkOn(chosenKey, ticksPerJob)
			// fmt.Printf("and will work on %s for %d ticks\n", string(chosenKey), workers[workerIndex].timeRemaining)
		}
		ticks++
	}
	return ticks - 1
}
